### PR TITLE
Patch cloudpickle to not reset dynamic class each time it is unpickled

### PR DESCRIPTION
**Pull request by sklam**
Fixes a problem with cloudpickle when dynamic class is unpickled and the class is reset to the state when it is pickled.

A short reproducer for the cloudpickle problem is available here:

```python
# Tested with cloudpickle 1.6.0
from cloudpickle import dumps, loads


class Klass:
    classvar = None

def mutator():
    Klass.classvar = 100

def check():
    print("checking....")
    print(f"   Klass.classvar [{hex(id(Klass))}] = {Klass.classvar}")


def failing_case():
    print("Klass", hex(id(Klass)))
    saved = dumps(Klass)
    mutator()
    check()
    loads(saved)
    check()
    loads(saved)
    check()



if __name__ == '__main__':
    failing_case()
```

Note, cloudpickle treats any class created in `__main__` as dynamic.

The patch makes cloudpickle remember whether a dynamic class is being reused. Reuse can come from the class being unpickled in the same process as it being pickled; or, a previously unpickled class is still around.

## Reference an existing issue

Fixes https://github.com/numba/numba/issues/7356

### CODE REVIEW

File: numba/tests/test_serialize.py [hunk 1]
```diff
@@ -205,5 +208,99 @@ def test_numba_unpickle(self):
         self.assertIs(got1, got2)
 
 
+
+class TestCloudPickleIssues(TestCase):
+    """This testcase includes all issues specific to cloudpickle implementation.
```
Comment (stuartarchibald) [inline numba/tests/test_serialize.py:213]:
```suggestion
    """This test case includes issues specific to the cloudpickle implementation.
```

File: numba/tests/test_serialize.py [hunk 2]
```diff
@@ -205,5 +208,99 @@ def test_numba_unpickle(self):
         self.assertIs(got1, got2)
 
 
+
+class TestCloudPickleIssues(TestCase):
```
Comment (stuartarchibald) [inline numba/tests/test_serialize.py:212]: Maybe this ought to also include the original reproducer as a test? https://github.com/numba/numba/issues/7356#issue-983961914

**Review by stuartarchibald — COMMENTED** [review]
Thanks for the patch, a couple of minor things to resolve else looks good.

File: numba/tests/test_serialize.py (hunk 2 above)
Comment (sklam) [reply numba/tests/test_serialize.py:213]: Good point. Test added in https://github.com/numba/numba/pull/7388/commits/8f6ffdcafbfdbb5317605950c4f57f43dae83111

**Review by stuartarchibald — APPROVED** [review]
Thanks for the patch and fixes.

**Pull request merged**